### DISCUSSION TITLE
Designing cross-platform component with platform-specific code

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#11838] [Q&A]
I'm currently working on refactoring [LibMpv.Client](https://github.com/homov/LibMpv) OpenGL integration into something useful.

Now it works fine on desktop, but for Android, I run into the issue that it requires platform-specific code.

[This](https://github.com/homov/LibMpv/blob/master/samples/Avalonia.LibMpv.Android.Sample.Android/App.axaml.cs#LL26C8-L26C8) must be called on startup, which requires a reference to `Avalonia.Android` and thus can't be in the main project.

[This](https://github.com/homov/LibMpv/blob/master/samples/Avalonia.LibMpv.Android.Sample.Android/ViewModels/MainViewModel.cs#L13) must be called when initializing the player context, only on Android.

[NativeVideoView](https://github.com/homov/LibMpv/blob/master/samples/Avalonia.LibMpv.Android.Sample.Android/Views/NativeVideoView.cs) is also platform-specific.

What's the right design to use here to embed platform-specific code into the library?

Since a reference to `Avalonia.Android` is required, it must go into a separate library. Is loading and using that library via reflection like a plugin the right thing to do?

On top of that, `NativeVideoView` is Avalonia-specific whereas everything else is platform-agnostic.

One approach would be to create all these libraries
- LibMpv
- LibMpv.Avalonia
- LibMpv.Avalonia.Android
- LibMpv.Android

That's kind of a mess.

Would love some feedback before delving too deep into this.

**Comment by mysteryx93** [comment]
Making some progress with the desktop implementation.
https://github.com/mysteryx93/LibMpv-OpenGL

It has 3 renderer implementations: Software, OpenGL and Native. If I initialize the sample app with either of the 3, it works. If I try to change the renderer at runtime, however, I'm getting weird behaviors.

You simply place a `MpvVideoView` on your page and set the `Renderer` property.
https://github.com/mysteryx93/LibMpv-OpenGL/blob/main/Src/LibMpv.Avalonia/MpvVideoView.cs

I then add another control as VisualChildren. Works the first time; not the 2nd time. After switching away and re-adding `SoftwareVideoView`, `Bounds` property has Width and Height set to 0, which is unusual, but GetVisualParent().Bounds gets the right values.

Also, if I add to LogicalChildren, or set Content of a ContentControl, the video displays in a new window instead of displaying within the window. It only works when I set as VisualChildren... and partially.

Is there something I should know about Avalonia behaviors with dynamically-added children?

**Reply by maxkatz6** [reply]
> Is there something I should know about Avalonia behaviors with dynamically-added children?

I, probably, would override AttachedToVisualTree and DettachedToVisualTree methods as well. But it will enforce typical controls lifetime on your library, i.e. disposing renderer on DettachedToVisualTree will also dispose of renderer when you switch TabControl tabs with these controls (as TabControl has only a single child at a time). 

Not sure what's the problem with bounds specifically. But I assume you should read bounds after layout was recalculated, for example with `Dispatcher.Post(Callback, Render)`.

**Reply by mysteryx93** [reply]
I try to access Bounds within Render which gets called repeatedly and it remains 0.  Tried to check with Dispatcher to make sure I'm on the right threat, no difference.

Adding to Control's VisualChildren, is that correct, and why is LogicalChildren and Content not working at all?

There is some very odd behavior here.

If you want to check, the demo is functional. [Clone the repo](https://github.com/mysteryx93/LibMpv-OpenGL), run `Sample.LibMpv.Avalonia.Desktop`. Click on Native and then back on Software.

When you click Load, the audio still plays, but without the video. In all 3 rendering modes.

But if Renderer is set before load, then all 3 work (on Linux).

**Reply by mysteryx93** [reply]
Maintaining state with TabControl is going to be a tough one; video would need to be reloaded with playback position restored. I don't think it's a good control to display in a TabControl.

I also might not want to reset when detaching from visual tree, as that will get triggered when switching from normal view to full screen mode.

**Comment by maxkatz6** [comment]
If you need to access platform-specific APIs, Microsoft didn't leave you a choice with a number of your libraries - there must be a special build for "net6.0-android" target. You simply can't access android APIs from "net6.0" unless you implement your own interop mappings, which I wouldn't recommend.
With or without reference to Avalonia.Android - that's up to your code, but I think it's necessary only for AndroidViewControlHandle.

> One approach would be to create all these libraries

It can only be "LibMpv" nuget package (with multiple targets per platform - net6.0, net6.0-android, net6.0-ios and possibly more) and "LibMpv.Avalonia" (the same with multiple targets). That's where Microsoft's idea of multi-targetting made things easier. **As you can have single project with multiple builds for multiple platforms.**

> [NativeVideoView](https://github.com/homov/LibMpv/blob/master/samples/Avalonia.LibMpv.Android.Sample.Android/Views/NativeVideoView.cs) is also platform-specific.

This is not necessary but depends on how it's easier for you.
I would recommend having a single shared control for all platforms, as users will most likely use it in shared XAML files. Native controls can be implementation details set up by internal code.

> [This](https://github.com/homov/LibMpv/blob/master/samples/Avalonia.LibMpv.Android.Sample.Android/App.axaml.cs#LL26C8-L26C8) must be called on startup

If you want to call some code on startup, you can write AppBuilder extension, as many first and third-party avalonia libraries do.
These extensions can be per platform (per project target/build)

Something like:
```
#if ANDROID // net6.0-android
public static AppBuilder UseLibMpv(this AppBuilder appBuilder)
{
   return appBuilder.AfterSetup(() => InitJVM());
}
#elif IOS // net6.0-ios
public static AppBuilder UseLibMpv(this AppBuilder appBuilder)
{
    // can be empty if nothing is required (yet?)
    return appBuilder;
}
#else // net6.0, i.e. most likely a desktop or browser
public static AppBuilder UseLibMpv(this AppBuilder appBuilder)
{
    // can be empty if nothing is required (yet?)
    return appBuilder;
}
#endif
```

Optionally these methods can have different names. 

And used in the user's code in platform-specific builder like here:
https://github.com/AvaloniaUI/avalonia-dotnet-templates/blob/master/templates/csharp/xplat/AvaloniaTest.Android/SplashActivity.cs#L18
Since it's a net6.0-android project, it will automatically pick up net6.0-android build.

**Reply by maxkatz6** [reply]
But please note, all of these recommendations are based on how MS does their platform specific libraries (primarily - MAUI). And not exactly how dotnet sdk team recommends it, as they recommend using runtime OS checks when possible - but it gets complicated pretty quickly. And since all android APIs are available only in net6.0-android target, we don't really have a choice but follow MAUI steps.

Also yet another complication is a browser, which doesn't have a dedicated target. [Yet](https://github.com/dotnet/designs/blob/main/accepted/2023/net8.0-browser-tfm.md).

**Reply by mysteryx93** [reply]
Wow thanks for the detailed feedback, this will give me a few options to work with!

I don't expect MPV to compile as WebAssembly anytime soon though. Although there *IS* a (slow) version of FFMPEG compiled as WebAssembly...
